local.go

```diff
@@ -54,59 +54,92 @@ func Local(f Function, initX []float64, settings *Settings, method Method) (*Res
 	}
 
 	startTime := time.Now()
-
 	funcs, funcInfo := getFunctionInfo(f)
 
-	if method == nil {
-		method = getDefaultMethod(funcInfo)
+	if settings.Recorder != nil {
+		// Initialize Recorder first. If it fails, we avoid the (possibly
+		// time-consuming) evaluation of F and DF at the starting location.
+		err := settings.Recorder.Init(funcInfo)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	if settings == nil {
 		settings = DefaultSettings()
 	}
-
 	stats := &Stats{}
-	location, err := getStartingLocation(funcs, funcInfo, initX, stats, settings)
+
+	optLoc, evalType, err := getStartingLocation(funcs, funcInfo, initX, stats, settings)
 	if err != nil {
 		return nil, err
 	}
-
-	optLoc := &Location{}
-	// update stats (grad norm, function value, etc.) so that things are
-	// initialized for the first convergence check
-	update(location, optLoc, stats, NoIteration, startTime)
-
+	// Send optLoc to Recorder before checking it for convergence.
 	if settings.Recorder != nil {
-		err = settings.Recorder.Init(funcInfo)
+		// Runtime is the only Stats field that needs to be updated here.
+		stats.Runtime = time.Since(startTime)
+		// TODO: Replace NoIteration with InitialIterate when it is added.
+		err = settings.Recorder.Record(optLoc, evalType, NoIteration, stats)
 		if err != nil {
-			return &Result{Status: Failure}, err
+			// Recorder failed, but at least we have the starting location.
+			// Build a meaningful result with Failure status and return it.
+			return &Result{*optLoc, *stats, Failure}, err
 		}
 	}
 
-	// actually perform the minimization
-	status, err := minimize(settings, location, method, funcInfo, stats, funcs, optLoc, startTime)
+	// Check if the starting location satisfies the convergence criteria.
+	stats.Runtime = time.Since(startTime)
+	// TODO: Replace MajorIteration with InitialIteration when it is added and
+	// checkConvergence() updated to handle it.
+	status := checkConvergence(optLoc, MajorIteration, stats, settings)
+	if status == NotTerminated {
+		// The starting location is not good enough, we need to start the
+		// minimization run.
+
+		// First, check the status of the function to see if it is able to
+		// evaluate.
+		// TODO: Is this the best place for checking the function status?
+		if funcInfo.IsStatuser {
+			// TODO: should the function really return status?
```

Comment:
I think so. It allows the function to signal completion for some reason. The only other way to communicate that is by falsely returning 0 for the gradient (or something). Two good reasons to have this:
1) The function can have an error for some reason. This is a way to put in that check without having to do it at every function evaluation where there is nothing the optimizer can do about it
2) In machine learning, one often wants to converge until the testing error is minimized (not the training error). The machine learning parameter function can return a status signifying early convergence even though a local minima has not been found.